metricbeat/beater/glob_change_detector.go

```diff
@@ -0,0 +1,79 @@
+package beater
+
+import (
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/mitchellh/hashstructure"
+)
+
+type GlobChangeDetector struct {
+	glob     string
+	lastScan time.Time
+	lastHash uint64
+}
+
+func NewGlobalChangeDetector(glob string) *GlobChangeDetector {
+	return &GlobChangeDetector{
+		lastScan: time.Time{},
+		lastHash: 0,
+		glob:     glob,
+	}
+}
+
+// Scan scans all files matched by the glob and checks if the number of files or the modtime of the files changed
+// It returns the list of files, a boolean if anything in the glob changed and potential errors.
+// To detect changes not only mod time is compared but also the hash of the files list. This is required to
+// also detect files which were removed.
+// The modtime is compared based on second as normally mod-time is in seconds. If it is unclear if something changed
+// the method will return true for the changes. It is strongly recommend to call scan not more frequent then 1s.
+func (gcd *GlobChangeDetector) Scan() ([]string, bool, error) {
+
+	globList, err := filepath.Glob(gcd.glob)
+	if err != nil {
+		return nil, false, err
+	}
+
+	updatedFiles := false
+	files := []string{}
+
+	lastScan := time.Now()
+	defer func() { gcd.lastScan = lastScan }()
+
+	for _, f := range globList {
+
+		info, err := os.Lstat(f)
+		if err != nil {
+			logp.Err("Error getting stats for file: %s", f)
+			continue
+		}
+
+		// Directories are symlinks are skipped
+		if !info.Mode().IsRegular() {
+			continue
+		}
+
+		// Check if one of the files was changed recently
+		// File modification time can be in seconds. -1 is to cover for files which
+		// were created during this second.
+		if info.ModTime().Sub(gcd.lastScan) >= -1*time.Second {
+			updatedFiles = true
+		}
+		files = append(files, f)
+	}
+
+	hash, err := hashstructure.Hash(files, nil)
+	if err != nil {
+		logp.Err("Error hashing files list: %s", err)
```

Comment:
I initially ignored the error as I didn't see an option how we ever get here an error. But to be sure we catch it in case it happens, I added the log line. 

If there is an error, the hash is 0. So in case hashing fails one time because of reason x, scan would still return changed which is the save way to go. I prefer to return changed once too many. If it continously fails, it will in the future depend on `updatedFiles` only. The not so nice part about this is that it does not detect removed files with updatedFiles.

Thinking about this again, probably the most consistent behaviour would be directly return `files, true, nil` in case of an error, to make sure in an error case `changed` is true. Will change it to this.